File: cathode-sync/src/main/java/net/simonvt/cathode/remote/sync/shows/SyncSeasons.java
/*
 * Copyright (C) 2014 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.simonvt.cathode.remote.sync.shows;

import android.database.Cursor;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import net.simonvt.cathode.api.entity.Season;
import net.simonvt.cathode.api.enumeration.Extended;
import net.simonvt.cathode.api.service.SeasonService;
import net.simonvt.cathode.common.database.Cursors;
import net.simonvt.cathode.jobqueue.JobPriority;
import net.simonvt.cathode.provider.DatabaseContract.SeasonColumns;
import net.simonvt.cathode.provider.ProviderSchematic.Seasons;
import net.simonvt.cathode.provider.helper.SeasonDatabaseHelper;
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper;
import net.simonvt.cathode.remote.CallJob;
import retrofit2.Call;

public class SyncSeasons extends CallJob<List<Season>> {

  @Inject transient SeasonService seasonService;

  @Inject transient ShowDatabaseHelper showHelper;
  @Inject transient SeasonDatabaseHelper seasonHelper;

  long traktId;

  public SyncSeasons(long traktId) {
    this.traktId = traktId;
  }

  @Override public String key() {
    return "SyncSeasons" + "&traktId=" + traktId;
  }

  @Override public int getPriority() {
    return JobPriority.SEASONS;
  }

  @Override public Call<List<Season>> getCall() {
    return seasonService.getSummary(traktId, Extended.FULL);
  }

  @Override public boolean handleResponse(List<Season> seasons) {
    final long showId = showHelper.getId(traktId);
    if (showId == -1L) {
      queue(new SyncShow(traktId));
      return true;
    }

    List<Long> seasonIds = new ArrayList<>();
    Cursor currentSeasons = getContentResolver().query(Seasons.fromShow(showId), new String[] {
        SeasonColumns.ID,
    }, null, null, null);
    while (currentSeasons.moveToNext()) {
      final long id = Cursors.getLong(currentSeasons, SeasonColumns.ID);
      seasonIds.add(id);
    }
    currentSeasons.close();

    for (Season season : seasons) {
      SeasonDatabaseHelper.IdResult result = seasonHelper.getIdOrCreate(showId, season.getNumber());
      seasonHelper.updateSeason(showId, season);
      seasonIds.remove(result.id);
      queue(new SyncSeason(traktId, season.getNumber()));
    }

    for (Long seasonId : seasonIds) {
      getContentResolver().delete(Seasons.withId(seasonId), null, null);
    }

    return true;
  }
}

File: cathode-sync/src/main/java/net/simonvt/cathode/actions/seasons/SyncSeasons.kt
/*
 * Copyright (C) 2014 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.simonvt.cathode.actions.seasons

import android.content.Context
import net.simonvt.cathode.actions.CallAction
import net.simonvt.cathode.actions.seasons.SyncSeasons.Params
import net.simonvt.cathode.api.entity.Season
import net.simonvt.cathode.api.enumeration.Extended
import net.simonvt.cathode.api.service.SeasonService
import net.simonvt.cathode.common.database.forEach
import net.simonvt.cathode.common.database.getLong
import net.simonvt.cathode.provider.DatabaseContract.SeasonColumns
import net.simonvt.cathode.provider.ProviderSchematic.Seasons
import net.simonvt.cathode.provider.delete
import net.simonvt.cathode.provider.helper.SeasonDatabaseHelper
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper
import net.simonvt.cathode.provider.query
import retrofit2.Call
import javax.inject.Inject

class SyncSeasons @Inject constructor(
  private val context: Context,
  private val seasonService: SeasonService,
  private val showHelper: ShowDatabaseHelper,
  private val seasonHelper: SeasonDatabaseHelper,
  private val syncSeason: SyncSeason
) : CallAction<Params, List<Season>>() {

  override fun key(params: Params): String = "SyncSeasons&traktId=${params.traktId}"

  override fun getCall(params: Params): Call<List<Season>> =
    seasonService.getSummary(params.traktId, Extended.FULL)

  override suspend fun handleResponse(params: Params, response: List<Season>) {
    val showId = showHelper.getId(params.traktId)
    val seasonIds = mutableListOf<Long>()
    val currentSeasons =
      context.contentResolver.query(Seasons.fromShow(showId), arrayOf(SeasonColumns.ID))
    currentSeasons.forEach { cursor -> seasonIds.add(cursor.getLong(SeasonColumns.ID)) }
    currentSeasons.close()

    for (season in response) {
      val result = seasonHelper.getIdOrCreate(showId, season.number)
      seasonHelper.updateSeason(showId, season)
      seasonIds.remove(result.id)
      syncSeason(SyncSeason.Params(params.traktId, season.number))
    }

    for (seasonId in seasonIds) {
      context.contentResolver.delete(Seasons.withId(seasonId))
    }
  }

  data class Params(val traktId: Long)
}
